socorro/processor/rules/mozilla.py

```diff
@@ -1097,3 +1097,47 @@ def action(self, raw_crash, dumps, processed_crash, tmpdir, status):
             item.strip() for item in utility_actors_name.split(",") if item.strip()
         ]
         processed_crash["utility_actors_name"] = names
+
+
+class ReportTypeRule(Rule):
+    """Determines the category for the report.
+
+    We get crash reports for different sorts of things. Some are from crashes. Others
+    are from browser hangs, content process hangs, shutdown hangs, etc. This determines
+    the type of the report and fills in the ``report_type`` field.
+
+    Bug #1667997
+
+    """
+
+    def identify_report_type(self, processed_crash, status):
+        # NOTE(willkg): This comes from the ipc_channel_error crash annotation
+        if "ipc_channel_error" in processed_crash:
+            return "hang"
+
+        # NOTE(willkg): This comes from the AsyncShutdownTimeout crash annotation
+        if "async_shutdown_timeout" in processed_crash:
+            return "hang"
+
+        # If this is a C++/Rust crash and there was a minidump and "RunWatchdog" is in
+        # the crashing thread stack, then it's a hang
+        if "json_dump" in processed_crash:
+            crash_data = processed_crash["json_dump"]
+            try:
+                crash_info = crash_data.get("crash_info") or {}
+                crashing_thread = int(crash_info.get("crashing_thread", 0))
+            except (TypeError, ValueError):
+                crashing_thread = 0
+
+            stack = glom(crash_data, "threads.%d.frames" % crashing_thread, default=[])
+            for frame in stack:
+                if "RunWatchdog" in (frame.get("function") or ""):
+                    return "hang"
```

Comment:
This matches the equivalent heuristics for the shutdown rule in signature generation. Because this rule runs before the signature generation rule in the processor pipeline, we can't use the results of signature generation to tell us it's a hang.

At some point, we might want to adjust signature generation to reflect the work done in processing rather than doing it's own thing.